examples/graph/gnn_citations.py

```diff
@@ -0,0 +1,728 @@
+                dropout=self.dropout_rate,
+                return_state=True,
+                recurrent_dropout=self.dropout_rate,
+            )
+        else:
+            self.update_fn = create_ffn(self.hidden_units, self.dropout_rate)
+
+    def prepare(self, node_repesentations, weights=None):
+        # node_repesentations shape is [num_edges, embedding_dim].
+        node_repesentations = self.ffn_prepare(node_repesentations)
+        if weights is not None:
+            node_repesentations = node_repesentations * tf.expand_dims(weights, -1)
+        return node_repesentations
+
+    def aggregate(self, node_indicies, neighbour_messages):
+        # node_indicies shape is [num_edges].
+        # neighbour_messages shape: [num_edges, representation_dim].
+        num_nodes = tf.math.reduce_max(node_indicies) + 1
+        if self.aggregation_type == "sum":
+            aggregated_message = tf.math.unsorted_segment_sum(
+                neighbour_messages, node_indicies, num_segments=num_nodes
+            )
+        elif self.aggregation_type == "mean":
+            aggregated_message = tf.math.unsorted_segment_mean(
+                neighbour_messages, node_indicies, num_segments=num_nodes
+            )
+        elif self.aggregation_type == "max":
+            aggregated_message = tf.math.unsorted_segment_max(
+                neighbour_messages, node_indicies, num_segments=num_nodes
+            )
+        else:
+            raise ValueError(f"Invalid aggregation type: {self.aggregation_type}.")
+
+        return aggregated_message
+
+    def update(self, node_repesentations, aggregated_messages):
+        # node_repesentations shape is [num_nodes, representation_dim].
+        # aggregated_messages shape is [num_nodes, representation_dim].
+        if self.combination_type == "gru":
+            # Create a sequence of two elements for the GRU layer.
+            h = tf.stack([node_repesentations, aggregated_messages], axis=1)
+        elif self.combination_type == "concat":
+            # Concatenate the node_repesentations and aggregated_messages.
+            h = tf.concat([node_repesentations, aggregated_messages], axis=1)
+        elif self.combination_type == "add":
+            # Add node_repesentations and aggregated_messages.
+            h = node_repesentations + aggregated_messages
+        else:
+            raise ValueError(f"Invalid combination type: {self.combination_type}.")
+
+        # Apply the processing function.
+        node_embeddings = self.update_fn(h)
+        if self.combination_type == "gru":
+            node_embeddings = tf.unstack(node_embeddings, axis=1)[-1]
+
+        if self.normalize:
+            node_embeddings = tf.nn.l2_normalize(node_embeddings, axis=-1)
+        return node_embeddings
+
+    def call(self, inputs):
+        # inputs is a tuple of three elements: node_repesentations, edges, edge_weights.
+        node_repesentations, edges, edge_weights = inputs
```

Comment:
Inputs should be documented in a docstring

Reply:
Done